cpp-ch/local-engine/Storages/Parquet/VectorizedParquetRecordReader.h

```diff
@@ -0,0 +1,254 @@
+
+protected:
+    std::unordered_set<int32_t> column_indices_;
+    const RowRanges & getRowRanges(int32_t row_group);
+    const ColumnIndexStore & getColumnIndexStore(int32_t row_group);
+
+    bool canPruningPage(const int32_t row_group) const { return column_index_filter_ && rowGroupPageIndexReader(row_group) != nullptr; }
+    std::unique_ptr<RowRanges> calculateRowRanges(const ColumnIndexStore & index_store, const size_t rowgroup_count) const
+    {
+        return std::make_unique<RowRanges>(column_index_filter_->calculateRowRanges(index_store, rowgroup_count));
+    }
+    std::unique_ptr<parquet::RowGroupMetaData> rowGroup(const int32_t row_group) const
+    {
+        const auto file_metadata = fileReader_->metadata();
+        return file_metadata->RowGroup(row_group);
+    }
+
+    std::shared_ptr<parquet::RowGroupPageIndexReader> rowGroupPageIndexReader(const int32_t row_group) const
+    {
+        const auto pageIndex = fileReader_->GetPageIndexReader();
+        return pageIndex == nullptr ? nullptr : pageIndex->RowGroup(row_group);
+    }
+
+    ColumnChunkPageRead
+    readColumnChunkPageBase(const parquet::RowGroupMetaData & rg, int32_t column_index, const BuildRead & build_read) const;
+
+public:
+    ParquetFileReaderExt(
+        const std::shared_ptr<::arrow::io::RandomAccessFile> & source,
+        std::unique_ptr<parquet::ParquetFileReader> parquetFileReader,
+        const std::shared_ptr<ColumnIndexFilter> & column_index_filter,
+        const std::vector<int32_t> & column_indices,
+        const DB::FormatSettings & format_settings);
+};
+
+class PageIterator final : public parquet::arrow::FileColumnIterator
+{
+    ParquetFileReaderExt * reader_ext_;
+
+public:
+    PageIterator(const int column_index, ParquetFileReaderExt * readerExt, const std::vector<int32_t> & row_groups)
+        : FileColumnIterator(column_index, readerExt->fileReader_.get(), row_groups), reader_ext_(readerExt)
+    {
+    }
+
+    ~PageIterator() override = default;
+
+    std::optional<ColumnChunkPageRead> nextChunkWithRowRange();
+};
+
+class VectorizedColumnReader
+{
+    std::shared_ptr<arrow::Field> arrowField_;
+    PageIterator input_;
+    std::shared_ptr<parquet::internal::RecordReader> record_reader_;
+    std::unique_ptr<ParquetReadState> read_state_;
+
+    void nextRowGroup();
+    void setPageReader(std::unique_ptr<parquet::PageReader> reader, const ReadSequence & read_sequence);
+
+public:
+    VectorizedColumnReader(
+        const parquet::arrow::SchemaField & field, ParquetFileReaderExt * reader, const std::vector<int32_t> & row_groups);
+    const std::string & col_name() const { return arrowField_->name(); }
+    bool hasMoreRead() const { return read_state_ && read_state_->hasMoreRead(); }
+    std::shared_ptr<arrow::ChunkedArray> readBatch(int64_t batch_size);
+};
+
+class VectorizedParquetRecordReader
+{
+    const DB::FormatSettings format_settings_;
+    DB::ArrowColumnToCHColumn arrowColumnToCHColumn_;
```

Comment:
using smart prt for DB::ArrowColumnToCHColumn doesn't have an explicitly declared default constructor.